persist/fs/index_lookup_test.go

```diff
@@ -77,6 +83,137 @@ func TestNewNearestIndexOffsetDetectsUnsortedFiles(t *testing.T) {
+	require.NoError(t, clone.close())
+}
+
+func TestClosingCloneDoesNotAffectParent(t *testing.T) {
+	indexSummaries := []schema.IndexSummary{
+		schema.IndexSummary{
+			Index:            0,
+			ID:               []byte("0"),
+			IndexEntryOffset: 0,
+		},
+		schema.IndexSummary{
+			Index:            1,
+			ID:               []byte("1"),
+			IndexEntryOffset: 10,
+		},
+	}
+
+	indexLookup := newIndexLookupWithValidData(t, indexSummaries)
+	clone, err := indexLookup.concurrentClone()
+	require.NoError(t, err)
+	require.NoError(t, clone.close())
+	for _, summary := range indexSummaries {
+		id := ts.StringID(string(summary.ID))
+		require.NoError(t, err)
+		offset, err := clone.getNearestIndexFileOffset(id)
+		require.NoError(t, err)
+		require.Equal(t, summary.IndexEntryOffset, offset)
+		id.Finalize()
+	}
+	require.NoError(t, indexLookup.close())
+}
+
+func TestParentAndClonesSafeForConcurrentUse(t *testing.T) {
+	numSummaries := 1000
+	numClones := 10
+
+	// Create test summary entries
+	indexSummaries := []schema.IndexSummary{}
+	for i := 0; i < numSummaries; i++ {
+		indexSummaries = append(indexSummaries, schema.IndexSummary{
+			Index:            int64(i),
+			ID:               []byte(strconv.Itoa(i)),
+			IndexEntryOffset: int64(10 * i),
+		})
+	}
+	sort.Sort(sortableSummaries(indexSummaries))
+
+	// Create indexLookup and associated clones
+	indexLookup := newIndexLookupWithValidData(t, indexSummaries)
+	clones := []*nearestIndexOffsetLookup{}
+	for i := 0; i < numClones; i++ {
+		clone, err := indexLookup.concurrentClone()
+		require.NoError(t, err)
+		clones = append(clones, clone)
+	}
+
+	// Spin up a goroutine for each clone that looks up every offset. Use one waitgroup
+	// to make a best effort attempt to get all the goroutines active before they start
+	// doing work, and then another waitgroup to wait for them to all finish their work.
+	startWg := sync.WaitGroup{}
+	doneWg := sync.WaitGroup{}
+	startWg.Add(len(clones) + 1)
+	doneWg.Add(len(clones) + 1)
+
+	lookupOffsetsFunc := func(clone *nearestIndexOffsetLookup) {
+		startWg.Done()
+		startWg.Wait()
+		for _, summary := range indexSummaries {
+			id := ts.StringID(string(summary.ID))
+			offset, err := clone.getNearestIndexFileOffset(id)
+			require.NoError(t, err)
+			require.Equal(t, summary.IndexEntryOffset, offset)
+			id.Finalize()
+		}
+		doneWg.Done()
+	}
+	go lookupOffsetsFunc(indexLookup)
+	for _, clone := range clones {
+		go lookupOffsetsFunc(clone)
+	}
+
+	// Wait for all workers to finish and then make sure everything can be cleaned
+	// up properly
+	doneWg.Wait()
+	require.NoError(t, indexLookup.close())
+	for _, clone := range clones {
+		require.NoError(t, clone.close())
+	}
+}
+
+func newIndexLookupWithValidData(t *testing.T, indexSummaries []schema.IndexSummary) *nearestIndexOffsetLookup {
+	// Create a temp file
+	file, err := ioutil.TempFile("", "index-lookup-sort")
+	require.NoError(t, err)
+	defer os.Remove(file.Name())
```

Comment:
Ha this works?  I assume its because the file descriptor is still active with the mmap, but it's funny that we delete the file but return a data-structure here that still has a ref to it (with the mmap) from the deleted file.

Probably fine to keep it this way, I just hope it's not OS dependent.

Reply:
you're right, I didn't realize what I was doing at the time haha. Cool that it works though